dbt/adapters/contracts/relation.py

```diff
@@ -18,13 +21,39 @@ class RelationType(StrEnum):
     Ephemeral = "ephemeral"
 
 
+class MaterializationContract(Protocol):
+    enforced: bool
+    alias_types: bool
+
+
+class MaterializationConfig(Mapping, ABC):
+    materialized: str
+    incremental_strategy: Optional[str]
+    persist_docs: Dict[str, Any]
+    column_types: Dict[str, Any]
+    full_refresh: Optional[bool]
+    quoting: Dict[str, Any]
+    unique_key: Union[str, List[str], None]
+    on_schema_change: Optional[str]
+    on_configuration_change: OnConfigurationChangeOption
+    contract: MaterializationContract
+    extra: Dict[str, Any]
+
+    def __contains__(self, item):
+        ...
+
+    def __delitem__(self, key):
+        ...
+
+
 class RelationConfig(Protocol):
     name: str
     database: str
     schema: str
     identifier: str
+    compiled_code: Optional[str]
     quoting_dict: Dict[str, bool]
-    config: Dict[str, str]
+    config: MaterializationConfig
```

Comment:
I don't think we should have a materialization config as an attribute of a relation config. Relations are components of a materialization, but the component matters (e.g. existing, target, temp, backup, etc.). Also, relations exist regardless of a materialization. So at very least, this needs to be optional.

Reply:
Yea this is confusing I was trying to figure out what RelationConfig.config actually represented and it seems to be the configuration for how an adapter should actually instantiate the resulting database object.

Maybe `RelationGenerationConfig`?

Reply:
Per our offline conversation, I'm fine leaving this as is.